src/rules/no-missing-label-refs.js

```diff
@@ -21,106 +21,67 @@ import { findOffsets, illegalShorthandTailPattern } from "../util.js";
-	// [foo][bar]
-	/\]\[([^\]]+)\]/u,
-
-	// [foo][]
-	/(\]\[\])/u,
-
-	// [foo]
-	/\[([^\]]+)\]/u,
-];
-
-const shorthandTailPattern = /\]\[\]$/u;
-
 /**
  * Finds missing references in a node.
  * @param {TextNode} node The node to check.
- * @param {string} docText The text of the node.
+ * @param {string} nodeText The text of the node.
  * @returns {Array<{label:string,position:Position}>} The missing references.
  */
-function findMissingReferences(node, docText) {
+function findMissingReferences(node, nodeText) {
 	const missing = [];
-	let startIndex = 0;
-	const offset = node.position.start.offset;
 	const nodeStartLine = node.position.start.line;
 	const nodeStartColumn = node.position.start.column;
 
-	/*
-	 * This loop works by searching the string inside the node for the next
-	 * label reference. If there is, it reports an error.
-	 * It then moves the start index to the end of the label reference and
-	 * continues searching the text until the end of the text is found.
-	 */
-	while (startIndex < node.value.length) {
-		const value = node.value.slice(startIndex);
-
-		const match = labelPatterns.reduce((previous, pattern) => {
-			if (previous) {
-				return previous;
-			}
-
-			return value.match(pattern);
-		}, null);
-
-		// check for array instead of null to appease TypeScript
-		if (!Array.isArray(match)) {
-			break;
-		}
+	const labelPattern = /\[(?<left>[^\]]*)\](?:\[(?<right>[^\]]*)\])?/dgu;
+	let match;
 
+	while ((match = labelPattern.exec(nodeText))) {
 		// skip illegal shorthand tail -- handled by no-invalid-label-refs
 		if (illegalShorthandTailPattern.test(match[0])) {
-			startIndex += match.index + match[0].length;
 			continue;
 		}
 
-		// Calculate the match index relative to just the node.
-		let columnStart = startIndex + match.index;
-		let label = match[1];
-
-		// need to look backward to get the label
-		if (shorthandTailPattern.test(match[0])) {
-			// adding 1 to the index just in case we're in a ![] and need to skip the !.
-			const startFrom = offset + startIndex + 1;
-			const lastOpenBracket = docText.lastIndexOf("[", startFrom);
-
-			if (lastOpenBracket === -1) {
-				startIndex += match.index + match[0].length;
-				continue;
-			}
-
-			label = docText
-				.slice(lastOpenBracket, match.index + match[0].length)
-				.match(/!?\[([^\]]+)\]/u)?.[1];
-			columnStart -= label.length;
-		} else if (match[0].startsWith("]")) {
-			columnStart += 2;
+		const { left, right } = match.groups;
+
+		// `[][]` or `[]`
+		if (!left && !right) {
+			continue;
+		}
+
+		let label, labelIndices;
+
+		if (right) {
+			label = right;
+			labelIndices = match.indices.groups.right;
 		} else {
-			columnStart += 1;
+			label = left;
+			labelIndices = match.indices.groups.left;
 		}
 
 		const { lineOffset: startLineOffset, columnOffset: startColumnOffset } =
-			findOffsets(node.value, columnStart);
-
-		const startLine = nodeStartLine + startLineOffset;
-		const startColumn = nodeStartColumn + startColumnOffset;
```

Comment:
Columns were wrong in some cases, for two reasons:

1. `TextNode#value` is not always the raw text of the node - some lines are trimmed left, and some aren't.
2. `findOffsets()`: `columnOffset` is the column offset only when it's on the same line. For other lines, `columnOffset` is 0-based column number, not offset.

Reply:
Good point on 2. Can you add comments to `findOffsets()` to document that?

Reply:
I added a note on how `columnOffset` should be used in https://github.com/eslint/markdown/pull/290/commits/63d3b4d05522a58a8aac468b794d41f8cf6fafe5.